lisa/transformers/kernel_installer.py

```diff
@@ -0,0 +1,308 @@
+        return "kernel_installer"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return KernelInstallerTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return []
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: KernelInstallerTransformerSchema = self.runbook
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node: RemoteNode = cast(
+            RemoteNode, Node.create(-1, runbook.connection, logger_name="kernel_node")
+        )
+        node.set_connection_info_by_runbook()
+        node.initialize()
+
+        uname = node.tools[Uname]
+        self._log.info(
+            f"kernel version before install: {uname.get_linux_information()}"
+        )
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
+        installer = factory.create_by_runbook(
+            runbook=runbook.installer, node=node, parent_log=self._log
+        )
+
+        installer.validate()
+        installer.install()
+
+        self._log.info("rebooting")
+        node.reboot()
+        self._log.info(
+            f"kernel version after install: "
+            f"{uname.get_linux_information(force_run=True)}"
+        )
+
+        return {}
+
+
+class RepoInstaller(BaseInstaller):
+    __menu_id_parts_pattern = re.compile(
+        r"^(?P<prefix>.*?)-.*-(?P<postfix>.*?-.*?-.*?-.*?-.*?-.*?)?$"
+    )
+
+    def __init__(
+        self,
+        runbook: Any,
+        node: RemoteNode,
+        parent_log: Logger,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(runbook, node, parent_log, *args, **kwargs)
+        self.repo_url = "http://archive.ubuntu.com/ubuntu/"
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "repo"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return RepoInstallerSchema
+
+    def validate(self) -> None:
+        assert isinstance(self._node.os, Ubuntu), (
+            f"The '{self.type_name()}' installer only support Ubuntu. "
+            f"The current os is {self._node.os.__class__.__name__}"
+        )
+
+    def install(self) -> None:
+        runbook: RepoInstallerSchema = self.runbook
+        node: RemoteNode = self._node
+        ubuntu: Ubuntu = cast(Ubuntu, node.os)
+        release = node.os.os_version.codename
+
+        assert (
+            release
+        ), f"cannot find codename from the os version: {node.os.os_version}"
+
+        # add the repo
+        if runbook.is_proposed:
+            version_name = f"{release}-proposed"
+        else:
+            version_name = release
+        repo_entry = (
+            f"deb {self.repo_url} {version_name} "
+            f"restricted main multiverse universe"
+        )
+        ubuntu.wait_running_package_process()
+        result = node.execute(f'add-apt-repository "{repo_entry}"', sudo=True)
```

Comment:
add-apt-repository -y

Reply:
What's the reason to add it? It runs well.